File: frontend/js/views/seller_disclosure.js
// Seller's property disclosure generator — known-defect statement with counts,
// via /calc/seller-disclosure. Previews live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
let LAST_DOC = null;

const STATUSES = [['no_issue', 'No known issue'], ['defect', 'Known defect'], ['unknown', 'Unknown']];
const SEED = [
    { cat: 'Roof', status: 'no_issue', note: '' },
    { cat: 'Plumbing', status: 'defect', note: 'leak under kitchen sink' },
    { cat: 'Foundation', status: 'unknown', note: 'never inspected' },
];

function statusOpts(sel) {
    return STATUSES.map(([v, l]) => `<option value="${v}" ${v === sel ? 'selected' : ''}>${esc(l)}</option>`).join('');
}

function rowHtml(it) {
    return `
        <div class="mpb-row ick-row">
            <input type="text" class="sd-cat" placeholder="${esc(t('view.sd.ph.cat'))}" value="${esc(it.cat || '')}">
            <select class="sd-status">${statusOpts(it.status || 'no_issue')}</select>
            <input type="text" class="sd-note" placeholder="${esc(t('view.sd.ph.note'))}" value="${esc(it.note || '')}">
            <button type="button" class="sd-del" data-i18n="view.sd.remove">Remove</button>
        </div>`;
}

export async function renderSellerDisclosure(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.sd.h1.title">// SELLER'S PROPERTY DISCLOSURE</span></h1>
        <p class="muted small" data-i18n="view.sd.hint.intro">
            The statement of known material defects a seller of real property gives the buyer before sale
            (required in most states). For each item the seller marks no known issue, a known defect, or
            unknown, with an explanation; the form counts the disclosed defects and unknowns. Drafting
            aid, not legal advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.sd.h2.inputs">Disclosure details</h2>
            <form id="sd-form" class="inline-form">
                <label><span data-i18n="view.sd.label.seller">Seller</span>
                    <input type="text" name="seller_name" value=""></label>
                <label><span data-i18n="view.sd.label.buyer">Buyer</span>
                    <input type="text" name="buyer_name" value=""></label>
                <label><span data-i18n="view.sd.label.property">Property address</span>
                    <input type="text" name="property_address" value=""></label>
                <label><span data-i18n="view.sd.label.date">Disclosure date</span>
                    <input type="date" name="disclosure_date" value="2026-06-15" required></label>
                <label><span data-i18n="view.sd.label.state">State (optional)</span>
                    <input type="text" name="state" value="California"></label>
                <label><span data-i18n="view.sd.label.asis">Sold as-is</span>
                    <input type="checkbox" name="as_is" checked></label>
            </form>
            <div class="mpb-head ick-head">
                <span data-i18n="view.sd.col.cat">Item</span>
                <span data-i18n="view.sd.col.status">Status</span>
                <span data-i18n="view.sd.col.note">Explanation</span>
                <span></span>
            </div>
            <div id="sd-rows">${SEED.map(rowHtml).join('')}</div>
            <button type="button" id="sd-add" class="secondary" data-i18n="view.sd.add">+ Add item</button>
        </div>
        <div id="sd-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#sd-form');
    const rowsEl = mount.querySelector('#sd-rows');

    const generate = async () => {
        const fd = new FormData(form);
        const items = [...rowsEl.querySelectorAll('.ick-row')].map((r) => ({
            category: (r.querySelector('.sd-cat').value || '').trim(),
            status: r.querySelector('.sd-status').value || 'no_issue',
            explanation: (r.querySelector('.sd-note').value || '').trim(),
        })).filter((x) => x.category);
        const body = {
            seller_name: (fd.get('seller_name') || '').trim(),
            buyer_name: (fd.get('buyer_name') || '').trim(),
            property_address: (fd.get('property_address') || '').trim(),
            items,
            as_is: fd.get('as_is') != null,
            disclosure_date: fd.get('disclosure_date'),
            state: (fd.get('state') || '').trim(),
        };
        try {
            const doc = await api.calcSellerDisclosure(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.sd.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);

    mount.querySelector('#sd-add').addEventListener('click', () => {
        rowsEl.insertAdjacentHTML('beforeend', rowHtml({ cat: '', status: 'no_issue', note: '' }));
        applyUiI18n(rowsEl.lastElementChild);
        generate();
    });
    rowsEl.addEventListener('click', (e) => {
        if (e.target.classList.contains('sd-del')) {
            e.target.closest('.ick-row').remove();
            generate();
        }
    });
    form.addEventListener('input', () => { live(); });
    rowsEl.addEventListener('input', () => { live(); });
    rowsEl.addEventListener('change', () => { live(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase(), ''];
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const el = mount.querySelector('#sd-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card ${doc.defect_count > 0 ? 'neg' : 'pos'}"><div class="label" data-i18n="view.sd.card.defects">Defects disclosed</div>
                    <div class="value">${doc.defect_count}</div></div>
                <div class="card"><div class="label" data-i18n="view.sd.card.unknown">Unknown</div>
                    <div class="value">${doc.unknown_count}</div></div>
                <div class="card"><div class="label" data-i18n="view.sd.card.items">Items</div>
                    <div class="value">${doc.item_count}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="sd-copy" type="button" data-i18n="view.sd.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="sd-download" type="button" data-i18n="view.sd.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#sd-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.sd.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.sd.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#sd-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'seller-disclosure.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/seller_disclosure.rs
//! Seller's property disclosure — the statement of known material defects a
//! seller of real property gives the buyer before sale (required in most
//! states). For each item the seller marks no known issue, a known defect, or
//! unknown, with an explanation; the form counts the disclosed defects and
//! unknowns. Distinct from the move-in/out inspection checklist (seller's
//! pre-sale liability, not tenancy condition). Drafting aid, not legal advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ItemStatus {
    NoIssue,
    Defect,
    Unknown,
}

impl ItemStatus {
    fn label(self) -> &'static str {
        match self {
            ItemStatus::NoIssue => "No known issue",
            ItemStatus::Defect => "KNOWN DEFECT",
            ItemStatus::Unknown => "Unknown",
        }
    }
}

#[derive(Debug, Clone, Deserialize)]
pub struct DisclosureItem {
    pub category: String,
    pub status: ItemStatus,
    #[serde(default)]
    pub explanation: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct SellerDisclosureInput {
    pub seller_name: String,
    #[serde(default)]
    pub buyer_name: String,
    pub property_address: String,
    #[serde(default)]
    pub items: Vec<DisclosureItem>,
    /// Sold as-is (the disclosure still applies to known defects).
    #[serde(default)]
    pub as_is: bool,
    pub disclosure_date: String,
    #[serde(default)]
    pub state: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct SellerDisclosure {
    pub title: String,
    pub item_count: usize,
    pub defect_count: usize,
    pub unknown_count: usize,
    pub clauses: Vec<DocClause>,
}

pub fn generate(i: &SellerDisclosureInput) -> SellerDisclosure {
    let defect_count = i.items.iter().filter(|x| x.status == ItemStatus::Defect).count();
    let unknown_count = i.items.iter().filter(|x| x.status == ItemStatus::Unknown).count();

    let disclosure_body = if i.items.is_empty() {
        "No items disclosed.".to_string()
    } else {
        i.items
            .iter()
            .map(|x| {
                let note = if x.explanation.trim().is_empty() {
                    String::new()
                } else {
                    format!(" — {}", x.explanation.trim())
                };
                format!("  • {}: {}{}", x.category, x.status.label(), note)
            })
            .collect::<Vec<_>>()
            .join("\n")
    };

    let summary_body = format!(
        "{} item(s) addressed; {} known defect(s) disclosed; {} marked unknown.",
        i.items.len(),
        defect_count,
        unknown_count
    );

    let as_is_body = if i.as_is {
        "The property is offered AS-IS. Selling as-is does NOT relieve the Seller of the duty to disclose known material defects; this disclosure is made in addition to the as-is sale.".to_string()
    } else {
        "The Seller makes this disclosure of known conditions; it is not a warranty and the Buyer is encouraged to obtain independent inspections.".to_string()
    };

    let mut clauses = vec![
        DocClause {
            heading: "Header".into(),
            body: format!(
                "Seller: {}\nBuyer: {}\nProperty: {}\nDisclosure date: {}",
                i.seller_name,
                if i.buyer_name.trim().is_empty() { "—" } else { i.buyer_name.trim() },
                i.property_address,
                i.disclosure_date
            ),
        },
        DocClause { heading: "1. Disclosure of Known Conditions".into(), body: disclosure_body },
        DocClause { heading: "2. Summary".into(), body: summary_body },
        DocClause { heading: "3. As-Is".into(), body: as_is_body },
        DocClause {
            heading: "4. Buyer Acknowledgment".into(),
            body: "The Buyer acknowledges receipt of this disclosure and understands it reflects the Seller's knowledge only, is not a substitute for the Buyer's own inspections, and does not warrant the condition of the property.".into(),
        },
    ];

    if !i.state.trim().is_empty() {
        clauses.push(DocClause {
            heading: "5. Governing Law".into(),
            body: format!("This disclosure is made under the law of the State of {}.", i.state.trim()),
        });
    }

    clauses.push(DocClause {
        heading: "Signatures".into(),
        body: format!(
            "Seller: ____________________  Date: {}\n{}\n\nBuyer: ____________________  Date: __________",
            i.disclosure_date, i.seller_name
        ),
    });

    SellerDisclosure {
        title: "Seller's Property Disclosure Statement".into(),
        item_count: i.items.len(),
        defect_count,
        unknown_count,
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn item(cat: &str, status: ItemStatus, note: &str) -> DisclosureItem {
        DisclosureItem { category: cat.into(), status, explanation: note.into() }
    }

    fn base() -> SellerDisclosureInput {
        SellerDisclosureInput {
            seller_name: "Sol Seller".into(),
            buyer_name: "Bea Buyer".into(),
            property_address: "100 Maple Ave".into(),
            items: vec![
                item("Roof", ItemStatus::NoIssue, ""),
                item("Plumbing", ItemStatus::Defect, "leak under kitchen sink"),
                item("Foundation", ItemStatus::Unknown, "never inspected"),
            ],
            as_is: true,
            disclosure_date: "2026-06-15".into(),
            state: "California".into(),
        }
    }

    #[test]
    fn counts() {
        let d = generate(&base());
        assert_eq!(d.item_count, 3);
        assert_eq!(d.defect_count, 1);
        assert_eq!(d.unknown_count, 1);
    }

    #[test]
    fn disclosure_lists_items_with_status() {
        let c = generate(&base()).clauses.into_iter().find(|c| c.heading.contains("Known Conditions")).unwrap();
        assert!(c.body.contains("Plumbing: KNOWN DEFECT — leak under kitchen sink"));
        assert!(c.body.contains("Roof: No known issue"));
        assert!(c.body.contains("Foundation: Unknown"));
    }

    #[test]
    fn summary_states_counts() {
        let c = generate(&base()).clauses.into_iter().find(|c| c.heading == "2. Summary").unwrap();
        assert!(c.body.contains("1 known defect(s) disclosed"));
        assert!(c.body.contains("1 marked unknown"));
    }

    #[test]
    fn as_is_still_requires_disclosure() {
        let c = generate(&base()).clauses.into_iter().find(|c| c.heading == "3. As-Is").unwrap();
        assert!(c.body.contains("does NOT relieve the Seller"));
    }

    #[test]
    fn not_as_is_message() {
        let c = generate(&SellerDisclosureInput { as_is: false, ..base() })
            .clauses.into_iter().find(|c| c.heading == "3. As-Is").unwrap();
        assert!(c.body.contains("not a warranty"));
    }

    #[test]
    fn empty_items() {
        let d = generate(&SellerDisclosureInput { items: vec![], ..base() });
        assert_eq!(d.defect_count, 0);
        let c = d.clauses.iter().find(|c| c.heading.contains("Known Conditions")).unwrap();
        assert!(c.body.contains("No items disclosed"));
    }

    #[test]
    fn governing_law_only_with_state() {
        assert!(generate(&base()).clauses.iter().any(|c| c.heading == "5. Governing Law"));
        let none = generate(&SellerDisclosureInput { state: String::new(), ..base() });
        assert!(!none.clauses.iter().any(|c| c.heading == "5. Governing Law"));
    }
}
